x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -273,12 +275,28 @@ public <A extends EsField> A readEsFieldWithCache() throws IOException {
                 return (A) esFieldFromCache(cacheId);
             }
         } else {
-            String className = readString();
+            String className = readCachedString();
             Writeable.Reader<? extends EsField> reader = EsField.getReader(className);
             return (A) reader.read(this);
         }
     }
 
+    public String readCachedString() throws IOException {
+        if (getTransportVersion().onOrAfter(TransportVersions.ESQL_CACHED_STRING_SERIALIZATION)) {
```

Comment:
I think we should have javadocs on these.

I'd prefer `if before(ES_CACHED_SER) { return readString(); }` without the `else`, but I'm I very "return early" programmer. It makes it easier for *me* to read these kinds of things. I'm not 100% sure it's easier on everyone else though.